demo/sample_azure_iot_embedded_sdk_pkcs11/common/demo_key_provisioning.c

```diff
@@ -0,0 +1,574 @@
+    xCertificateTemplate.xSubject.type = CKA_SUBJECT;
+    xCertificateTemplate.xSubject.pValue = xSubject;
+    xCertificateTemplate.xSubject.ulValueLen = strlen( ( const char * ) xSubject );
+    xCertificateTemplate.xValue.type = CKA_VALUE;
+    xCertificateTemplate.xValue.pValue = ( CK_VOID_PTR ) pucCertificate;
+    xCertificateTemplate.xValue.ulValueLen = ( CK_ULONG ) xCertificateLength;
+    xCertificateTemplate.xLabel.type = CKA_LABEL;
+    xCertificateTemplate.xLabel.pValue = ( CK_VOID_PTR ) pucLabel;
+    xCertificateTemplate.xLabel.ulValueLen = strlen( ( const char * ) pucLabel );
+    xCertificateTemplate.xCertificateType.type = CKA_CERTIFICATE_TYPE;
+    xCertificateTemplate.xCertificateType.pValue = &xCertificateType;
+    xCertificateTemplate.xCertificateType.ulValueLen = sizeof( CK_CERTIFICATE_TYPE );
+    xCertificateTemplate.xTokenObject.type = CKA_TOKEN;
+    xCertificateTemplate.xTokenObject.pValue = &xTokenStorage;
+    xCertificateTemplate.xTokenObject.ulValueLen = sizeof( xTokenStorage );
+
+    xResult = C_GetFunctionList( &pxFunctionList );
+
+    /* Test for a valid certificate: 0x2d is '-', as in ----- BEGIN CERTIFICATE. */
+    if( ( pucCertificate == NULL ) || ( pucCertificate[ 0 ] != 0x2d ) )
+    {
+        xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+    }
+
+    if( xResult == CKR_OK )
+    {
+        /* Convert the certificate to DER format if it was in PEM. The DER key
+         * should be about 3/4 the size of the PEM key, so mallocing the PEM key
+         * size is sufficient. */
+        pucDerObject = pvPortMalloc( xCertificateTemplate.xValue.ulValueLen );
+        xDerLen = xCertificateTemplate.xValue.ulValueLen;
+
+        if( pucDerObject != NULL )
+        {
+            lConversionReturn = convert_pem_to_der( xCertificateTemplate.xValue.pValue,
+                                                    xCertificateTemplate.xValue.ulValueLen,
+                                                    pucDerObject,
+                                                    &xDerLen );
+
+            if( 0 != lConversionReturn )
+            {
+                xResult = CKR_ARGUMENTS_BAD;
+            }
+        }
+        else
+        {
+            xResult = CKR_HOST_MEMORY;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        /* Set the template pointers to refer to the DER converted objects. */
+        xCertificateTemplate.xValue.pValue = pucDerObject;
+        xCertificateTemplate.xValue.ulValueLen = xDerLen;
+    }
+
+    /* Best effort clean-up of the existing object, if it exists. */
+    if( xResult == CKR_OK )
+    {
+        xDestroyProvidedObjects( xSession,
+                                 &pucLabel,
+                                 &xCertificateClass,
+                                 1 );
+    }
+
+    /* Create an object using the encoded client certificate. */
+    if( xResult == CKR_OK )
+    {
+        configPRINTF( ( "Write certificate...\r\n" ) );
```

Comment:
Are we moving logging to `configPRINTF` base?  (Sorry been out of loop a bunch.)  If we are this is cool, if not I'd want to understand why this isn't the 'AZ_LOG` or whatever we had?